source/core/slang-crypto.cpp

```diff
@@ -0,0 +1,603 @@
+/*
+ * MD5 implementation is based on:
+ * http://openwall.info/wiki/people/solar/software/public-domain-source-code/md5
+ * Original file header is at the bottom of this file.
+ * 
+ * SHA1 implementation is based on:
+ * https://github.com/983/SHA1
+ * Original LICENSE is at the bottom of this file.
+ */
+
+#include "slang-crypto.h"
+#include "../core/slang-char-util.h"
+
+namespace Slang
+{
+
+// DigestUtil
+
+/*static*/ String DigestUtil::digestToString(const void* digest, SlangInt digestSize)
+{
+    SLANG_ASSERT(digest && digestSize >= 0);
+
+    static const char* hex = "0123456789abcdef";
+
+    String str;
+    const uint8_t* data = reinterpret_cast<const uint8_t*>(digest);
+    for (SlangInt i = 0; i < digestSize; ++i)
+    {
+        str.append(hex[data[i] >> 4]);
+        str.append(hex[data[i] & 0xf]);
+    }
+    return str;
+}
+
+/*static*/ bool DigestUtil::stringToDigest(const char* str, SlangInt strLength, void *digest, SlangInt digestSize)
+{
+    SLANG_ASSERT(str && strLength >= 0 && digest && digestSize >= 0);
+
+    if (strLength != digestSize * 2)
+    {
+        ::memset(digest, 0, digestSize);
+        return false;
+    }
+
+    uint8_t* data = reinterpret_cast<uint8_t*>(digest);
+    for (SlangInt i = 0; i < digestSize; ++i)
+    {
+        int upper = CharUtil::getHexDigitValue(str[i * 2]);
+        int lower = CharUtil::getHexDigitValue(str[i * 2 + 1]);
+        if (upper == -1 || lower == -1)
+        {
+            ::memset(digest, 0, digestSize);
+            return false;
+        }
+        data[i] = uint8_t(lower | upper << 4);;
+    }
+
+    return true;
+}
+
+// MD5
+
+MD5::MD5()
+{
+    init();
+}
+
+void MD5::init()
+{
+    m_lo = 0;
+    m_hi = 0;
+    m_a = 0x67452301;
+    m_b = 0xefcdab89;
+    m_c = 0x98badcfe;
+    m_d = 0x10325476;
+}
+
+void MD5::update(const void* data, SlangInt size)
+{
+    uint32_t saved_lo;
+    SlangInt used, available;
+
+    saved_lo = m_lo;
+    if ((m_lo = (saved_lo + size) & 0x1fffffff) < saved_lo)
+        m_hi++;
+    m_hi += (uint32_t)size >> 29;
+
+    used = saved_lo & 0x3f;
+
+    if (used) {
```

Comment:
Nitpik - Allman is the preferred indenting style.

Reply:
Yeah, that code was copied over. I'll do a cleanup pass with the next PR.